### PR TITLE
feature/paywall-use-case

### PR DESCRIPTION
**Pull request by TheUnderScorer**

### CODE REVIEW

**Review by ilfa — APPROVED** [review]

File: api/paywall/article-views.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+      },
+      timestamp: {
+        [Op.between]: [timestampStart, timestampEnd],
+      },
+    },
+  });
+}
+
+/**
+ * Checks if given visitor has exceeded his daily limit of free article views.
```
Comment (makma) [inline api/paywall/article-views.js:60]:
```suggestion
 * Checks if the given visitor has exceeded his daily limit of free article views.
```

File: api/paywall/article-views.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
+import { Op } from 'sequelize';
+import { ArticleView } from './database';
+import { CheckResult, checkResultType, messageSeverity } from '../../shared/server';
+import { getTodayDateRange } from '../../shared/utils/date';
+import { ARTICLE_VIEW_LIMIT } from '../../shared/paywall/constants';
+
+/**
+ * Saves article view into database. If it already exists, we update its timestamp.
```
Comment (makma) [inline api/paywall/article-views.js:8]:
```suggestion
 * Saves article view into the database. If it already exists, we update its timestamp.
```

File: api/paywall/articles.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
+/**
+ * Mock articles "database". In a real world, these articles could be fetched from anywhere.
```
Comment (makma) [inline api/paywall/articles.js:2]:
```suggestion
 * Mock articles "database". In the real world, these articles could be fetched from anywhere.
```

File: api/paywall/paywall-endpoint.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  ensureValidRequestIdAndVisitorId,
+  getForbiddenReponse,
+  getVisitorData,
+  reportSuspiciousActivity,
+} from '../../shared/server';
+import { initPaywall } from './database';
+import { checkCountOfViewedArticles } from './article-views';
+
+// Base checks for every endpoint related to loan risk
+const loanChecks = [
```
Comment (makma) [inline api/paywall/paywall-endpoint.js:17]: probably not `loanChecks` + comment

File: api/paywall/paywall-endpoint.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  checkFreshIdentificationRequest,
+  checkConfidenceScore,
+  checkIpAddressIntegrity,
+  checkOriginsIntegrity,
+  // Additional checks that returns error if user exceeded his daily free article views
+  checkCountOfViewedArticles,
+];
+
+// Provides common logic used in paywall use-case
```
Comment (makma) [inline api/paywall/paywall-endpoint.js:26]:
```suggestion
// Provides common logic used in the paywall use-case
```

File: api/paywall/paywall-endpoint.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    for (const ruleCheck of [...loanChecks, ...optionalChecks]) {
+      const result = await ruleCheck(visitorData, req);
+
+      if (result) {
+        switch (result.type) {
+          case checkResultType.Passed:
+          case checkResultType.Challenged:
+            continue;
+          case checkResultType.ArticleViewLimitExceeded:
+            // No need to report suspicious activity for this error
```
Comment (makma) [inline api/paywall/paywall-endpoint.js:58]: Good catch and comment!

File: pages/api/paywall/article/[id].js [hunk 7]
```diff
@@ -0,0 +1,32 @@
+import { articles } from '../../../../api/paywall/articles';
+import { paywallEndpoint } from '../../../../api/paywall/paywall-endpoint';
+import { countViewedArticles, saveArticleView } from '../../../../api/paywall/article-views';
+import { ARTICLE_VIEW_LIMIT } from '../../../../shared/paywall/constants';
+
+/**
+ * Fetches article by its ID. Supports paywall logic, which means that we keep track how many articles were viewed by given user.
```
Comment (makma) [inline pages/api/paywall/article/[id].js:7]:
```suggestion
 * Fetches article by its ID. Supports paywall logic, which means that we keep track of how many articles were viewed by a given user.
```

File: pages/api/paywall/article/[id].js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+ * If user has exceeded limit of articles that he can view for free, we return an error.
```
Comment (makma) [inline pages/api/paywall/article/[id].js:8]:
```suggestion
 * If a user has exceeded limit of articles that he can view for free, we return an error.
```

File: pages/api/paywall/get-articles.js [hunk 9]
```diff
@@ -0,0 +1,20 @@
+import { articles } from '../../../api/paywall/articles';
+
+const SHORT_ARTICLE_CONTENT_LENGTH = 100;
+
+/**
+ * Returns articles with trimmed description from the database.
```
Comment (makma) [inline pages/api/paywall/get-articles.js:6]:
```suggestion
 * Returns articles with trimmed descriptions from the database.
```

**Review by makma — CHANGES_REQUESTED** [review]
Great work @Przemo, just a couple of minor things to adjust, thank you! :)
- Change count of articles from 3 to 2 (fewer steps for users in the demo, better and easier to understand the showcase)
- Use some real content instead of lorem ipsum e.g. from https://fingerprint.com/blog/. Just a couple of first paragraphs should be sufficient. It's hard to remember visited articles with lorem ipsum, therefore more real-world content should be IMO better. Alternatively, you can use http://bullshitipsum.com/. Up to you.
- An error after viewing an article after limitation just floats in the middle of the screen. Do we have another alternative? This looks weird to me...
- `This page demonstrates paywall implementation using Fingerprint PRO`. -> It's `Fingerprint Pro`, not `Fingerprint PRO`
- Add personalization use case and paywall use case to the readme.md
- Remove warning from the readme.
- Add a link to the readme referencing our content https://fingerprint.com/use-cases/

**Review by makma — APPROVED** [review]
Thank you @TheUnderScorer!

**Pull request merged**